src/mui/input/ReferenceManyInput.js

```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    setPagination = (pagination) => {
+        if (pagination !== this.param.pagination) {
+            this.param.pagination = pagination;
+            this.fetchReferenceAndOptions();
+        }
+    }
+
+    setSort = (sort) => {
+        if (sort !== this.params.sort) {
+            this.params.sort = sort;
+            this.fetchReferenceAndOptions();
+        }
+    }
+
+    fetchReferenceAndOptions({ input, reference, source, resource } = this.props) {
+        const { pagination, sort, filter } = this.params;
+        const ids = input.value;
+        if (ids) {
+            if (!Array.isArray(ids)) {
+                throw Error("Value of ReferenceManyInput should be an array");
+            }
+            this.props.crudGetMany(reference, ids);
+        }
+        this.props.crudGetMatching(reference, referenceSource(resource, source), pagination, sort, filter);
+    }
+
+    render() {
+        const { input, resource, label, source, referenceRecords, allowEmpty, matchingReferences, basePath, onChange, children, meta } = this.props;
+        if (!(referenceRecords && referenceRecords.length > 0) && !allowEmpty) {
+            return <Labeled
+                label={typeof label === 'undefined' ? `resources.${resource}.fields.${source}` : label}
+                source={source}
+                resource={resource}
+            />;
+        }
+
+        return React.cloneElement(children, {
+            allowEmpty,
+            input,
+            label: typeof label === 'undefined' ? `resources.${resource}.fields.${source}` : label,
+            resource,
+            meta,
+            source,
+            choices: matchingReferences,
+            basePath,
+            onChange,
+            filter: noFilter, // for AutocompleteInput
+            setFilter: this.debouncedSetFilter,
+            setPagination: this.setPagination,
+            setSort: this.setSort,
+        });
+    }
+}
+
+ReferenceManyInput.propTypes = {
+    addField: PropTypes.bool.isRequired,
+    allowEmpty: PropTypes.bool.isRequired,
+    basePath: PropTypes.string,
+    children: PropTypes.element.isRequired,
+    crudGetMatching: PropTypes.func.isRequired,
+    crudGetMany: PropTypes.func.isRequired,
+    filter: PropTypes.object,
+    filterToQuery: PropTypes.func.isRequired,
+    input: PropTypes.object.isRequired,
+    label: PropTypes.string,
+    matchingReferences: PropTypes.array,
+    meta: PropTypes.object,
+    onChange: PropTypes.func,
+    perPage: PropTypes.number,
+    reference: PropTypes.string.isRequired,
+    referenceRecords: PropTypes.array,
+    resource: PropTypes.string.isRequired,
+    sort: PropTypes.shape({
+        field: PropTypes.string,
+        order: PropTypes.oneOf(['ASC', 'DESC']),
+    }),
+    source: PropTypes.string,
+};
+
+ReferenceManyInput.defaultProps = {
+    allowEmpty: false,
+    filter: {},
+    filterToQuery: searchText => ({ q: searchText }),
+    matchingReferences: [],
+    meta: {},
```

Comment:
I've removed this one from other components because it has side effects, please remove it here, too.